### PR TITLE
proposal: use errors table and code generation

### PR DESCRIPTION
**Pull request by daniel-pape**
WIP (not ready to merge)

Proposal for issue "Custom Error type #37”.

Idea would be to use a small database to centrally store errors, codes and documentation and generate the code (still as it was) from this data. Potential benefits:

* Table is more easy to scan and handle
* Separates data (error) and their representation code (can generate different code if desired)
* Provides additional uniqueness checks for codes

### CODE REVIEW

**Comment by spetz** [comment]
Hey Daniel,

Thanks a lot for the contribution, it looks great! Funny thing is that I've just made a simple improvement to error mappings by using [phf](https://github.com/rust-phf/rust-phf) in #63 like 30 minutes ago :D

Your approach looks way better! I was wondering if it would be possible to avoid the SQL dependency and use the [sled.rs](http://sled.rs) instead, which we already use in the few places, like storing the users' data etc.? It's a very simple to use embedded DB, and we could just stick to the single storage type. Just a thought, what do you think?

**Comment by hubcio** [comment]
> Hey Daniel,
> 
> Thanks a lot for the contribution, it looks great! Funny thing is that I've just made a simple improvement to error mappings by using [phf](https://github.com/rust-phf/rust-phf) in #63 like 30 minutes ago :D
> 
> Your approach looks way better! I was wondering if it would be possible to avoid the SQL dependency and use the [sled.rs](http://sled.rs) instead, which we already use in the few places, like storing the users' data etc.? It's a very simple to use embedded DB, and we could just stick to the single storage type. Just a thought, what do you think?

Great code @daniel-pape, thank you for contribution! Can you make it work with `sled`? We would like to avoid two databases dependency during compilation.

**Comment by daniel-pape** [comment]
> Hey Daniel,
> 
> Thanks a lot for the contribution, it looks great! Funny thing is that I've just made a simple improvement to error mappings by using [phf](https://github.com/rust-phf/rust-phf) in #63 like 30 minutes ago :D
> 
> Your approach looks way better! I was wondering if it would be possible to avoid the SQL dependency and use the [sled.rs](http://sled.rs) instead, which we already use in the few places, like storing the users' data etc.? It's a very simple to use embedded DB, and we could just stick to the single storage type. Just a thought, what do you think?

Thanks for the fast and welcoming feedback @spetz and @BartoszCiesla! I am not familiar with `sled` and will give it try. After a first brief look I would imagine that handling constraints is not possible with it (as opposed to say SQLite). But I totally agree that one should try to keep the number of dependencies as low as possible. So let me take a look at it and get back to you in the next days.

**Comment by spetz** [comment]
Yes, lack of the unique constraint would be a tradeoff, but I guess that in this case, some additional tests validating the overall uniqueness of all error codes could do the trick. Overall, everything sounds good, and feel free to ping us at any time if you'd need some help with this! Also, don't worry if the #63 would be merged in the meantime, and just feel free to get rid of this code :)

File: iggy/build.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        .join("")
+}
+
+fn get_spaces(num_spaces: usize) -> String {
+    if num_spaces == 0 {
+        "".to_string()
+    } else {
+        " ".repeat(num_spaces).to_string()
+    }
+}
```
Comment (hubcio) [inline iggy/build.rs:21]: you can simplify it:
```rust
fn get_spaces(num_spaces: usize) -> String {
    " ".repeat(num_spaces)
}
```

**Review by hubcio — COMMENTED** [review]
I will continue review tomorrow :) but overall looks good

File: iggy/build.rs (hunk 1 above)
Comment (daniel-pape) [reply iggy/build.rs:21]: Good point! Changed it.

**Comment by spetz** [comment]
Hey, I've just checked the code and successfully executed the tests :)

To make it compile, I only had to fix the listed conflicting file (simply replace the `error.rs` contents with `pub use crate::errors::*;`), and also include 2 new errors that were recently added to the `master` branch:

```rust
ErrorRepositoryEntry {
    snake_case_name: "invalid_jwt_algorithm".to_string(),
    code: 63,
    signature: "String".to_string(),
    converts_from: "".to_string(),
    template: "Invalid JWT algorithm: {0}".to_string(),
},
ErrorRepositoryEntry {
    snake_case_name: "invalid_jwt_secret".to_string(),
    code: 64,
    signature: "".to_string(),
    converts_from: "".to_string(),
    template: "Invalid JWT secret".to_string(),
},
```

If you could just update the code with these, to allow the automatic merge that'd be great :)

I also ran (as usual) `cargo clippy --all-targets` and got a few minor suggestions about the possible improvements - could you take a look at these, and then we will be very close to merging this feature? :)

Speaking of the code, it works as expected and this contribution will definitely improve the way we deal with the errors, so once again, thank you for this!

**Comment by spetz** [comment]
Hey, I've just checked the code and successfully executed the tests :)

To make it compile, I only had to fix the listed conflicting file and also include 2 new errors that were recently added to the `master` branch:

```rust
ErrorRepositoryEntry {
    snake_case_name: "invalid_jwt_algorithm".to_string(),
    code: 63,
    signature: "String".to_string(),
    converts_from: "".to_string(),
    template: "Invalid JWT algorithm: {0}".to_string(),
},
ErrorRepositoryEntry {
    snake_case_name: "invalid_jwt_secret".to_string(),
    code: 64,
    signature: "".to_string(),
    converts_from: "".to_string(),
    template: "Invalid JWT secret".to_string(),
},
```

If you could just update the code with these, to allow the automatic merge that'd be great :)

I also ran (as usual) `cargo clippy --all-targets` and got a few minor suggestions about the possible improvements - could you take a look at these, and then we will be very close to merging this feature? :)

Speaking of the code, it works as expected and this contribution will definitely improve the way we deal with the errors, so once again, thank you for this!

File: iggy/build.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+use errors_repository::PreprocessedErrorRepositoryEntry;
+
+use crate::data_repository::{DataRepository, SledDb};
+use crate::errors_repository::get_or_create;
+
+mod data_repository;
+mod errors_repository;
+
+fn snake_to_pascal_case(s: &str) -> String {
```
Comment (hubcio) [inline iggy/build.rs]: if it fits you, just use this crate:
https://docs.rs/convert_case/latest/convert_case/enum.Case.html

less boilerplate code :)

File: iggy/build.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        ];
+
+        lines.join("\n")
+    }
+}
+
+const GENERATED_ERRORS_PATH: &'static str = "./src/errors/generated_code/errors.rs";
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let errors_db: SledDb = get_or_create().expect("Could not load errors database");
```
Comment (hubcio) [inline iggy/build.rs]: stick to one way of handling errors: either use expect() everywhere, or return with syntax sugar `?` like you did at line 206. I prefer second option.

**Review by hubcio — CHANGES_REQUESTED** [review]
One more thing: call `rustfmt` on code after generation. Overall, I like your code.

File: iggy/build.rs (hunk 2 above)
Comment (daniel-pape) [reply iggy/build.rs]: Changed it.

File: iggy/build.rs (hunk 3 above)
Comment (daniel-pape) [reply iggy/build.rs]: Updated

**Comment by daniel-pape** [comment]
> Hey, I've just checked the code and successfully executed the tests :)
> 
> To make it compile, I only had to fix the listed conflicting file and also include 2 new errors that were recently added to the `master` branch:
> 
> ```rust
> ErrorRepositoryEntry {
>     snake_case_name: "invalid_jwt_algorithm".to_string(),
>     code: 63,
>     signature: "String".to_string(),
>     converts_from: "".to_string(),
>     template: "Invalid JWT algorithm: {0}".to_string(),
> },
> ErrorRepositoryEntry {
>     snake_case_name: "invalid_jwt_secret".to_string(),
>     code: 64,
>     signature: "".to_string(),
>     converts_from: "".to_string(),
>     template: "Invalid JWT secret".to_string(),
> },
> ```
> 
> If you could just update the code with these, to allow the automatic merge that'd be great :)
> 
> I also ran (as usual) `cargo clippy --all-targets` and got a few minor suggestions about the possible improvements - could you take a look at these, and then we will be very close to merging this feature? :)
> 
> Speaking of the code, it works as expected and this contribution will definitely improve the way we deal with the errors, so once again, thank you for this!

Updated correspondingly

File: iggy/build.rs (hunk 3 above)
Comment (daniel-pape) [reply iggy/build.rs]: Also added the rustfmt step

**Review by hubcio — APPROVED** [review]
LGTM

**Comment by hubcio** [comment]
btw, don't worry about squashing, it'll be done automatically upon merge by gh actions bot

**Comment by hubcio** [comment]
@daniel-pape  can you rebase with master? we'll merge after you do it :)

**Comment by hubcio** [comment]
still something's wrong, are you rebasing against our master (from iggy-rs:master, not your fork master)?
image

**Review by spetz — APPROVED** [review]

**Pull request merged**